test/e2e_node/node_problem_detector_linux.go

```diff
@@ -237,9 +233,19 @@ var _ = framework.KubeDescribe("NodeProblemDetector [NodeFeature:NodeProblemDete
 					},
 					Containers: []v1.Container{
 						{
-							Name:    name,
-							Image:   image,
-							Command: []string{"sh", "-c", "touch " + logFile + " && /node-problem-detector --logtostderr --system-log-monitors=" + configFile + fmt.Sprintf(" --apiserver-override=%s?inClusterConfig=true", framework.TestContext.Host)},
+							Name:  name,
+							Image: image,
+							Command: []string{
+								"sh",
+								"-c",
+								// `ServiceAccount` admission controller is disabled in node e2e tests, so we could not use
+								// inClusterConfig here.
+								fmt.Sprintf("touch %s && /node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override=%s?inClusterConfig=false&auth=%s",
```

Comment:
It took me some time, but in the end, the reason why it failed with no error and logs...
When you execute the command with the `sh -c` you should be aware of special characters, in your case you have `--apiserver-override=%s?inClusterConfig=false&auth=%s"` that includes `&` that interpreted as a special character, so the final command has view
```
touch %s && \
/node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override=%s?inClusterConfig=false & \
auth=%s
```
when `&` bitwise and, `auth=some_variable` always have zero output, so the final value will be 0(at least I think so, did not check it).

To make it work you just need to escape `&` -> `\&`

Reply:
That's a possibility. I tested it (`\&` is not a valid escape sequence):

```
$ git diff
diff --git a/test/e2e_node/node_problem_detector_linux.go b/test/e2e_node/node_problem_detector_linux.go
index 6750a2ceee5..e91046d9434 100644
--- a/test/e2e_node/node_problem_detector_linux.go
+++ b/test/e2e_node/node_problem_detector_linux.go
@@ -240,7 +240,7 @@ current-context: local-context
                                                                "-c",
                                                                // `ServiceAccount` admission controller is disabled in node e2e tests, so we could not use
                                                                // inClusterConfig here.
-                                                               fmt.Sprintf("touch %s && /node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override=%s?inClusterConfig=false&auth=%s",
+                                                               fmt.Sprintf(`touch %s && /node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override=%s?inClusterConfig=false&auth=%s`,
                                                                        logFile,
                                                                        configFile,
                                                                        framework.TestContext.Host,
```

The test still fails - same symptoms - even though the docker entrypoint looks correct:

```
            "Entrypoint": [
                "sh",
                "-c",
                "touch /log/test.log && /node-problem-detector --logtostderr --system-log-monitors=/config/testconfig.json --apiserver-override=https://127.0.0.1:6443?inClusterConfig=false&auth=/config/kubeconfig"
            ],
```

Then I tried again with:

```
$ git diff
diff --git a/test/e2e_node/node_problem_detector_linux.go b/test/e2e_node/node_problem_detector_linux.go
index 6750a2ceee5..3eca64593e9 100644
--- a/test/e2e_node/node_problem_detector_linux.go
+++ b/test/e2e_node/node_problem_detector_linux.go
@@ -231,20 +231,36 @@ current-context: local-context
                                                        },
                                                },
                                        },
-                                       Containers: []v1.Container{
+                                       InitContainers: []v1.Container{
                                                {
-                                                       Name:  name,
-                                                       Image: image,
-                                                       Command: []string{
-                                                               "sh",
+                                                       Name:    "init-log-file",
+                                                       Image:   "debian",
+                                                       Command: []string{"/bin/sh"},
+                                                       Args: []string{
                                                                "-c",
-                                                               // `ServiceAccount` admission controller is disabled in node e2e tests, so we could not use
-                                                               // inClusterConfig here.
-                                                               fmt.Sprintf("touch %s && /node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override=%s?inClusterConfig=false&auth=%s",
-                                                                       logFile,
-                                                                       configFile,
-                                                                       framework.TestContext.Host,
-                                                                       kubeConfigFile),
+                                                               fmt.Sprintf("touch %s", logFile),
+                                                       },
+                                                       VolumeMounts: []v1.VolumeMount{
+                                                               {
+                                                                       Name:      logVolume,
+                                                                       MountPath: path.Dir(logFile),
+                                                               },
+                                                               {
+                                                                       Name:      localtimeVolume,
+                                                                       MountPath: etcLocaltime,
+                                                               },
+                                                       },
+                                               },
+                                       },
+                                       Containers: []v1.Container{
+                                               {
+                                                       Name:    name,
+                                                       Image:   image,
+                                                       Command: []string{"/node-problem-detector"},
+                                                       Args: []string{
+                                                               "--logtostderr",
+                                                               fmt.Sprintf("--system-log-monitors=%s", configFile),
+                                                               fmt.Sprintf("--apiserver-override=%s?inClusterConfig=false&auth=%s", framework.TestContext.Host, kubeConfigFile),
                                                        },
                                                        Env: []v1.EnvVar{
                                                                {
```

This works fine:

```
Ran 1 of 330 Specs in 112.085 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 329 Skipped
```

The docker command/args look like this:

```
        "Path": "/node-problem-detector",
        "Args": [
            "--logtostderr",
            "--system-log-monitors=/config/testconfig.json",
            "--apiserver-override=https://127.0.0.1:6443?inClusterConfig=false&auth=/config/kubeconfig"
        ],
```

So I don't think the escape is the issue (or at least the way I'm escaping).

Reply:
> When you execute the command with the `sh -c` you should be aware of special characters

@cynepco3hahue Ah I think this make sense! Because the `node-problem-detector` was put in background, so the container exit immediately and we got no error and logs. Thanks for enlightening us.

Reply:
@karan I think the point is sh would treat `&` in a special way, perhaps you could try:
```
`touch %s && /node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override='%s?inClusterConfig=false&auth=%s'`
```
note the single quote around the value of `--apiserver-override`

Reply:
Yep exactly, I verified that your PR with
```
fmt.Sprintf("touch %s && /node-problem-detector --logtostderr --system-log-monitors=%s --apiserver-override=%s?inClusterConfig=false\\&auth=%s",
									logFile,
									configFile,
									framework.TestContext.Host,
									kubeConfigFile),
```
Take in account that you need double slash it `\\&`:)

Reply:
Ah gotcha! @knight42 feel free to undo my changes in your PR in lieu of the escaping here.

Reply:
@karan I think your change is beneficial which makes the command easier to read, I'd like to keep it.